src/tof/pulse.py

```diff
@@ -51,116 +52,226 @@ def __init__(
 
         Parameters
         ----------
+        kind:
+            Name of a pre-defined pulse from a neutron facility.
+        tmin:
+            Start time of the pulse.
+        tmax:
+            End time of the pulse.
+        wmin:
+            Minimum wavelength of the pulse.
+        wmax:
+            Maximum wavelength of the pulse.
         neutrons:
             Number of neutrons in the pulse.
+        sampling:
+            Number of points used to sample the probability distributions.
         """
-        if neutrons is not None:
-            self.neutrons = neutrons
-        if self.kind is not None:
-            params = getattr(facilities, self.kind)
-            if self.tmin is None:
-                self.tmin = params.time.coords['time'].min()
-            if self.tmax is None:
-                self.tmax = params.time.coords['time'].max()
-            if self.lmin is None:
-                self.lmin = params.wavelength.coords['wavelength'].min()
-            if self.lmax is None:
-                self.lmax = params.wavelength.coords['wavelength'].max()
-            self.tmin = self.tmin.to(unit='s')
-            self.tmax = self.tmax.to(unit='s')
-            self.lmin = self.lmin.to(unit='angstrom')
-            self.lmax = self.lmax.to(unit='angstrom')
-
-            x_time = np.linspace(
-                self.tmin.value, self.tmax.value, self.sampling_resolution
-            )
-            x_wav = np.linspace(
-                self.lmin.value, self.lmax.value, self.sampling_resolution
-            )
+        pulse = cls(
+            tmin=tmin,
+            tmax=tmax,
+            wmin=wmin,
+            wmax=wmax,
+            neutrons=neutrons,
+            generate=False,
+        )
+        pulse.kind = kind
+        params = getattr(facilities, pulse.kind)
+        if pulse.tmin is None:
+            pulse.tmin = params.time.coords['time'].min()
+        if pulse.tmax is None:
+            pulse.tmax = params.time.coords['time'].max()
+        if pulse.wmin is None:
+            pulse.wmin = params.wavelength.coords['wavelength'].min()
+        if pulse.wmax is None:
+            pulse.wmax = params.wavelength.coords['wavelength'].max()
+        pulse.tmin = pulse.tmin.to(unit='s')
+        pulse.tmax = pulse.tmax.to(unit='s')
+        pulse.wmin = pulse.wmin.to(unit='angstrom')
+        pulse.wmax = pulse.wmax.to(unit='angstrom')
+
+        x_time = np.linspace(pulse.tmin.value, pulse.tmax.value, sampling)
+        x_wav = np.linspace(pulse.wmin.value, pulse.wmax.value, sampling)
+        p_time = np.interp(
+            x_time,
+            params.time.coords['time'].to(unit='s').values,
+            params.time.values,
+        )
+        p_time /= p_time.sum()
+        p_wav = np.interp(
+            x_wav,
+            params.wavelength.coords['wavelength'].to(unit='angstrom').values,
+            params.wavelength.values,
+        )
+        p_wav /= p_wav.sum()
```

Comment:
Have you considered using https://scipp.github.io/generated/modules/scipy/scipp.scipy.interpolate.interp1d.html?highlight=interp1d?

Reply:
I think originally i went with `numpy` because it was all `numpy` internally, no `scipp`, and I didn't want to add `scipy` as a dependency. But now we have `scipp` and also depend on `scipy` for the constants, so I will try it and see if it makes things cleaner.